I believe some error snuck into the code in the last edit of the URL parsing. I can't load the integration properly in my virtual test environment. See stacktrace below. Does everything work ok in your setup @spiddeer ?


> 2025-10-29 21:56:01.246 ERROR (MainThread) [aiohttp.server] Error handling request from 172.17.0.1
> Traceback (most recent call last):
>   File "/home/vscode/.local/ha-venv/lib/python3.13/site-packages/aiohttp/web_protocol.py", line 510, in _handle_request
>     resp = await request_handler(request)
>            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/home/vscode/.local/ha-venv/lib/python3.13/site-packages/aiohttp/web_app.py", line 569, in _handle
>     return await handler(request)
>            ^^^^^^^^^^^^^^^^^^^^^^
>   File "/home/vscode/.local/ha-venv/lib/python3.13/site-packages/aiohttp/web_middlewares.py", line 117, in impl
>     return await handler(request)
>            ^^^^^^^^^^^^^^^^^^^^^^
>   File "/workspaces/home-assistant-core/homeassistant/components/http/security_filter.py", line 92, in security_filter_middleware
>     return await handler(request)
>            ^^^^^^^^^^^^^^^^^^^^^^
>   File "/workspaces/home-assistant-core/homeassistant/components/http/forwarded.py", line 87, in forwarded_middleware
>     return await handler(request)
>            ^^^^^^^^^^^^^^^^^^^^^^
>   File "/workspaces/home-assistant-core/homeassistant/components/http/request_context.py", line 26, in request_context_middleware
>     return await handler(request)
>            ^^^^^^^^^^^^^^^^^^^^^^
>   File "/workspaces/home-assistant-core/homeassistant/components/http/ban.py", line 86, in ban_middleware
>     return await handler(request)
>            ^^^^^^^^^^^^^^^^^^^^^^
>   File "/workspaces/home-assistant-core/homeassistant/components/http/auth.py", line 242, in auth_middleware
>     return await handler(request)
>            ^^^^^^^^^^^^^^^^^^^^^^
>   File "/workspaces/home-assistant-core/homeassistant/components/http/headers.py", line 41, in headers_middleware
>     response = await handler(request)
>                ^^^^^^^^^^^^^^^^^^^^^^
>   File "/workspaces/home-assistant-core/homeassistant/helpers/http.py", line 73, in handle
>     result = await handler(request, **request.match_info)
>              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/workspaces/home-assistant-core/homeassistant/components/http/decorators.py", line 83, in with_admin
>     return await func(self, request, *args, **kwargs)
>            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/workspaces/home-assistant-core/homeassistant/components/http/data_validator.py", line 74, in wrapper
>     return await method(view, request, data, *args, **kwargs)
>            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/workspaces/home-assistant-core/homeassistant/components/config/config_entries.py", line 188, in post
>     return await self._post_impl(request, data)
>            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/workspaces/home-assistant-core/homeassistant/components/config/config_entries.py", line 195, in _post_impl
>     return await super()._post_impl(request, data)
>            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/workspaces/home-assistant-core/homeassistant/helpers/data_entry_flow.py", line 92, in _post_impl
>     result = self._prepare_result_json(result)
>   File "/workspaces/home-assistant-core/homeassistant/components/config/config_entries.py", line 215, in _prepare_result_json
>     return _prepare_config_flow_result_json(result, super()._prepare_result_json)
>   File "/workspaces/home-assistant-core/homeassistant/components/config/config_entries.py", line 154, in _prepare_config_flow_result_json
>     return prepare_result_json(result)
>   File "/workspaces/home-assistant-core/homeassistant/helpers/data_entry_flow.py", line 52, in _prepare_result_json
>     data["data_schema"] = voluptuous_serialize.convert(
>                           ~~~~~~~~~~~~~~~~~~~~~~~~~~~~^
>         schema, custom_serializer=cv.custom_serializer
>         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>     )
>     ^
>   File "/home/vscode/.local/ha-venv/lib/python3.13/site-packages/voluptuous_serialize/__init__.py", line 59, in convert
>     pval = convert(value, custom_serializer=custom_serializer)
>   File "/home/vscode/.local/ha-venv/lib/python3.13/site-packages/voluptuous_serialize/__init__.py", line 164, in convert
>     raise ValueError(f"Unable to convert schema: {schema}")
> ValueError: Unable to convert schema: <function Url at 0x707ab280fe20>

--- a/custom_components/infometric/config_flow.py
+++ b/custom_components/infometric/config_flow.py
@@ -28,42 +28,49 @@ class ConfigFlow(config_entries.ConfigFlow, domain=DOMAIN):
         errors: dict[str, str] = {}
 
         if user_input is not None:
-            # Normalize URL using urlparse
-            parsed_url = urlparse(user_input[CONF_URL])
-            normalized_url = f"{parsed_url.scheme}://{parsed_url.netloc}{parsed_url.path.rstrip('/')}"
-            
-            # Uniqueness check: same URL + username
-            for entry in self._async_current_entries():
-                entry_parsed = urlparse(entry.data.get(CONF_URL, ""))
-                entry_normalized = f"{entry_parsed.scheme}://{entry_parsed.netloc}{entry_parsed.path.rstrip('/')}"
-                if (
-                    entry_normalized == normalized_url
-                    and entry.data.get(CONF_USERNAME) == user_input[CONF_USERNAME]
-                ):
-                    return self.async_abort(reason="already_configured")
-
-            client = InfometricClient(
-                normalized_url,
-                user_input[CONF_USERNAME],
-                user_input[CONF_PASSWORD],
-            )
+            # Validate and normalize URL using urlparse
             try:
-                await client.authenticate(
-                    aiohttp_client.async_get_clientsession(self.hass)
-                )
-            except Exception as auth_err:  # Differentiate later if needed
-                errors["base"] = "auth_failed"
-            else:
-                user_input[CONF_URL] = normalized_url
-                return self.async_create_entry(
-                    title=user_input.get(CONF_NAME, DEFAULT_NAME), data=user_input
-                )
+                parsed_url = urlparse(user_input[CONF_URL])
+                if not parsed_url.scheme or not parsed_url.netloc:
+                    errors["base"] = "invalid_url"
+                else:
+                    normalized_url = f"{parsed_url.scheme}://{parsed_url.netloc}{parsed_url.path.rstrip('/')}"
+                    
+                    # Uniqueness check: same URL + username
+                    for entry in self._async_current_entries():
+                        entry_parsed = urlparse(entry.data.get(CONF_URL, ""))
+                        entry_normalized = f"{entry_parsed.scheme}://{entry_parsed.netloc}{entry_parsed.path.rstrip('/')}"
+                        if (
+                            entry_normalized == normalized_url
+                            and entry.data.get(CONF_USERNAME) == user_input[CONF_USERNAME]
+                        ):
+                            return self.async_abort(reason="already_configured")
+
+                    if not errors:
+                        client = InfometricClient(
+                            normalized_url,
+                            user_input[CONF_USERNAME],
+                            user_input[CONF_PASSWORD],
+                        )
+                        try:
+                            await client.authenticate(
+                                aiohttp_client.async_get_clientsession(self.hass)
+                            )
+                        except Exception as auth_err:  # Differentiate later if needed
+                            errors["base"] = "auth_failed"
+                        else:
+                            user_input[CONF_URL] = normalized_url
+                            return self.async_create_entry(
+                                title=user_input.get(CONF_NAME, DEFAULT_NAME), data=user_input
+                            )
+            except Exception:
+                errors["base"] = "invalid_url"
 
         return self.async_show_form(
             step_id="user",
             data_schema=vol.Schema(
                 {
-                    vol.Required(CONF_URL, default=f"{DEFAULT_URL}/"): vol.Url(),
+                    vol.Required(CONF_URL, default=f"{DEFAULT_URL}/"): str,
                     vol.Required(CONF_USERNAME): str,
                     vol.Required(CONF_PASSWORD): str,
                     vol.Optional(CONF_NAME, default=DEFAULT_NAME): str,

--- a/custom_components/infometric/diagnostics.py
+++ b/custom_components/infometric/diagnostics.py
@@ -1,0 +1,52 @@
+"""Diagnostics support for Infometric Panorama integration."""
+from __future__ import annotations
+
+from typing import Any
+
+from homeassistant.components.diagnostics import async_redact_data
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import CONF_PASSWORD, CONF_USERNAME
+from homeassistant.core import HomeAssistant
+
+from .const import DOMAIN
+
+TO_REDACT = {CONF_PASSWORD, CONF_USERNAME}
+
+async def async_get_config_entry_diagnostics(
+    hass: HomeAssistant, entry: ConfigEntry
+) -> dict[str, Any]:
+    """Return diagnostics for a config entry."""
+    coordinator = hass.data[DOMAIN].get(entry.entry_id)
+
+    data: dict[str, Any] = {
+        "entry": {
+            "title": entry.title,
+            "data": async_redact_data(entry.data, TO_REDACT),
+            "options": entry.options,
+        }
+    }
+
+    if coordinator:
+        data["coordinator"] = {
+            "last_update_success": getattr(coordinator, "last_update_success", None),
+            "last_update_time": getattr(coordinator, "last_update_time", None).isoformat()
+            if getattr(coordinator, "last_update_time", None)
+            else None,
+        }
+        if getattr(coordinator, "data", None):
+            meters = []
+            for meter_type in ["energy", "hotwater", "coldwater"]:
+                meter = getattr(coordinator.data, meter_type, None)
+                if meter:
+                    meters.append(
+                        {
+                            "type": meter_type,
+                            "id": meter.id,
+                            "total": meter.total,
+                            "monthly_average": meter.monthly_average,
+                            "monthly_prognosis": meter.monthly_prognosis,
+                        }
+                    )
+            data["meters"] = meters
+
+    return data

--- a/custom_components/infometric/infometric.py
+++ b/custom_components/infometric/infometric.py
@@ -45,15 +45,16 @@ class InfometricClient:
         with hidden token parsing if the site introduces anti-forgery tokens.
         """
         self._session = session
-        login_payload = (
-            f"UserName={self._username}&Password={self._password}&"
-            "RememberMe=true&commit=Logga+in"
-        )
+        login_payload = {
+            "UserName": self._username,
+            "Password": self._password,
+            "RememberMe": "true",
+            "commit": "Logga in",
+        }
         try:
             resp = await self._session.post(
                 url=self._base_url,
                 data=login_payload,
-                headers={"content-type": "application/x-www-form-urlencoded"},
             )
         except ClientError as err:
             raise InfometricException(f"Network error during authenticate: {err}") from err

--- a/custom_components/infometric/manifest.json
+++ b/custom_components/infometric/manifest.json
@@ -7,5 +7,5 @@
   "iot_class": "cloud_polling",
   "issue_tracker": "https://github.com/sillymoi/homeassistant-infometric/issues",
   "requirements": [],
-  "version": "0.0.3"
+  "version": "0.0.4"
 }

